### PR TITLE
boards: arm: GD32F450I-EVAL: add initial support

### PR DESCRIPTION
**Pull request by gmarull**
This PR adds initial support for the GD32F450I-EVAL board. It also includes some refactoring to the `hal_gigadevice` module and SoC/dts organization to accommodate addition of new series. Each commit contains more information/details on the changes performed.

Note: I'll open a separate PR for the `pinctrl` driver, for now, the board has a custom init function.

### CODE REVIEW

File: dts/arm/gigadevice/gd32f4xx/gd32f450ik.dtsi [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	soc {
+		flash-controller@40023c00 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(3072)>;
+			};
+		};
+
+		sram0: memory@20000000 {
+			reg = <0x20000000 DT_SIZE_K(112)>;
+		};
```
Comment (nandojve) [inline dts/arm/gigadevice/gd32f4xx/gd32f450ik.dtsi:20]: This is same for all 4xx series and can be defined at gd32f4xx.dtsi. Only SRAM2 and ADDSRAM should be defined here for f450.

File: dts/arm/gigadevice/gd32f4xx/gd32f4xx.dtsi [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+				reg = <0xe000ed90 0x40>;
+				arm,num-mpu-regions = <8>;
+			};
+		};
+	};
+
+	soc {
+		sram0: memory@20000000 {
+			compatible = "mmio-sram";
+		};
```
Comment (nandojve) [inline dts/arm/gigadevice/gd32f4xx/gd32f4xx.dtsi:32]: SRAM0, SRAM1, DTCM and BKP SRAM are same for all F4XX series and should be define here

File: soc/arm/gigadevice/gd32f4xx/soc.h [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * @file SoC configuration header for the GD32F4XX SoC series.
+ */
+
+#ifndef _SOC_ARM_GIGADEVICE_GD32F4XX_SOC_H_
+#define _SOC_ARM_GIGADEVICE_GD32F4XX_SOC_H_
+
+#include <devicetree.h>
+#include <gd32f4xx.h>
```
Comment (nandojve) [inline soc/arm/gigadevice/gd32f4xx/soc.h:14]:
```
#include <sys/util.h>

#ifndef _ASMLANGUAGE

#include <devicetree.h>
#include <gd32f4xx.h>

#endif /* !_ASMLANGUAGE */
```

File: boards/arm/gd32f450i_eval/gd32f450i_eval.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2021, Teslabs Engineering S.L.
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: gd32f450i_eval
+name: GigaDevice GD32F450I Evaluation Kit
+type: mcu
+arch: arm
+ram: 256
```
Comment (nandojve) [inline boards/arm/gd32f450i_eval/gd32f450i_eval.yaml:8]: SRAM0 doesn't have this value.

File: boards/arm/gd32f450i_eval/support/openocd.cfg [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2021, Teslabs Engineering S.L.
```
Comment (nandojve) [inline boards/arm/gd32f450i_eval/support/openocd.cfg:1]: I'll ask to add ATL-Electronics rights too for this specific file.

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @gmarull ,

Just a few comments.

File: boards/arm/gd32f450i_eval/support/openocd.cfg (hunk 5 above)
Comment (gmarull) [reply boards/arm/gd32f450i_eval/support/openocd.cfg:2]: adjusted

File: boards/arm/gd32f450i_eval/gd32f450i_eval.yaml (hunk 4 above)
Comment (gmarull) [reply boards/arm/gd32f450i_eval/gd32f450i_eval.yaml]: adjusted to sram0 value

File: soc/arm/gigadevice/gd32f4xx/soc.h (hunk 3 above)
Comment (gmarull) [reply soc/arm/gigadevice/gd32f4xx/soc.h:12]: added ASM guards

File: dts/arm/gigadevice/gd32f4xx/gd32f4xx.dtsi (hunk 2 above)
Comment (gmarull) [reply dts/arm/gigadevice/gd32f4xx/gd32f4xx.dtsi]: Moved sram0 to gd32f4xx.dtsi file, other rams have not defined (some are not supported yet, others likely have specific purpose again unsupported)

File: dts/arm/gigadevice/gd32f4xx/gd32f450ik.dtsi (hunk 1 above)
Comment (gmarull) [reply dts/arm/gigadevice/gd32f4xx/gd32f450ik.dtsi]: done

**Review by nandojve — APPROVED** [review]
LGTM !

**Review by cameled — APPROVED** [review]
LGTM.

File: soc/arm/gigadevice/gd32f4xx/Kconfig.series [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2021, Teslabs Engineering S.L.
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_GD32F4XX
+	bool "GigaDevice GD32F4XX series Cortex-M4F MCU"
+	select ARM
+	select CPU_HAS_ARM_MPU
+	select CPU_HAS_FPU
+	select CPU_CORTEX_M4
+	select CPU_CORTEX_M_HAS_SYSTICK
```
Comment (cameled) [inline soc/arm/gigadevice/gd32f4xx/Kconfig.series:10]: CPU_CORTEX_M_HAS_SYSTICK  and CPU_CORTEX_M_HAS_VTOR config are not required here, CPU_CORTEX_M4 select them.

Comment (nandojve) [reply soc/arm/gigadevice/gd32f4xx/Kconfig.series:10]: +1

Comment (gmarull) [reply soc/arm/gigadevice/gd32f4xx/Kconfig.series]: removed, thanks!

**Review by carlescufi — APPROVED** [review]
LGTM

**Review by martinjaeger — APPROVED** [review]
PR looks good. Haven't worked with this particular MCU series, though, and I didn't double-check with the reference manual.

**Pull request merged**